Adding psycopg2 integration

hectorhdzg opened the pull request:
Took OpenCensus implementation as starting point for this one, we store more properties in traces now and we use tracer provided during  trace_integration method

https://github.com/census-instrumentation/opencensus-python/tree/master/contrib/opencensus-ext-postgresql

psycopg docs
http://initd.org/psycopg/docs/index.html 

#265

hectorhdzg commented on ext/opentelemetry-ext-postgresql/tests/test_postgresql_integration.py:
```diff
@@ -0,0 +1,31 @@
+from opentelemetry import trace as trace_api
+
+
+class TestPostgresqlIntegration(unittest.TestCase):
```
Still working on unit tests, will add soon

Oberon00 commented on ext/opentelemetry-ext-postgresql/src/opentelemetry/ext/postgresql/__init__.py:
```diff
@@ -0,0 +1,94 @@
+
+    wrapt.wrap_function_wrapper(psycopg2, "connect", wrap)
+
+    class TraceCursor(pgcursor):
```
Why doesn't this use opentelemetry-ext-dbapi from #264? Did you run into the issue I described in https://github.com/open-telemetry/opentelemetry-python/pull/264/files#r348452864? In any case, instead of deriving from pgcursor, you might want to use a wrapt ObjectProxy https://wrapt.readthedocs.io/en/latest/wrappers.html#object-proxy. That way, you could implement a generic solution. Or you could add support for such cursor_factory instrumentations to ext-dbapi.

Another question: What happens if the user manually specifies/sets cursor_factory? Will either our or the user's cursor_factory be ignored?

hectorhdzg replied:
I based this in current OpenCensus implementation and started working with this before adding dbapi in other PR, makes sense to reuse it as much as possible, cursor_factory in fact will be overriden like you mentioned and is current behavior in OpenCensus, will file a bug for it and ensure it doesn't happen here

hectorhdzg replied:
cursor_factory is a not thing in the Python DB API specification, so I guess it doesn't makes sense to add it in dbapi ext, dbapi ext is currently wrapping connect method, the actual connection and the cursor, in psycopg all of these are implemented in C, that is why we are hooking up to the cursor_factory, so I'm still not sure how we can reuse dbapi ext here, any suggestion would be appreciated.

Using ObjectProxy instead of pgcursor sounds like a good approach

toumorokoshi replied:
I think this could be an extension of the DBAPI PR. looking at the chunk that actually starts a span and sets values, it's very similar to the DBApi functionality.

Although one caveat as a con is the fact that the API is slightly different in that psycopg2 accepts sql.SQL objects rather than just strings, but that could reconciled with a light wrapper on all of them.

ultimately this is an auto-instrumentation though, so as long as we don't expose APIs that people may rely on, we can probably feel safe marking this as a potential refactor.

hectorhdzg replied:
Updated the code to use dbapi ext as much as possible, including span attributes related to connection population and actual span creation functionality

condorcet commented on ext/opentelemetry-ext-postgresql/src/opentelemetry/ext/postgresql/__init__.py:
```diff
@@ -0,0 +1,94 @@
+            if database:
+                name += "." + database
+
+            query = args[0] if args else ""
```
Please note that query in psycopg2 might be `Composable` object
> The objects exposed by the sql module can be used to compose a query as a Python string (using the as_string() method) or passed directly to cursor methods such as execute(), executemany(), copy_expert().
http://initd.org/psycopg/docs/sql.html

We have a problem with it in OT instrumentation once https://github.com/uber-common/opentracing-python-instrumentation/pull/82

hectorhdzg commented:
I would like to reuse dbapi package being added here https://github.com/open-telemetry/opentelemetry-python/pull/264, I will update this PR once is available

toumorokoshi commented on ext/opentelemetry-ext-postgresql/setup.cfg:
```diff
@@ -0,0 +1,47 @@
+# limitations under the License.
+#
+[metadata]
+name = opentelemetry-ext-postgresql
```
is this the right name for the package? I guess there's a few different DBAPI options available: https://docs.sqlalchemy.org/en/13/dialects/postgresql.html

hectorhdzg replied:
Updated to psycopg2

toumorokoshi commented on ext/opentelemetry-ext-postgresql/src/opentelemetry/ext/postgresql/__init__.py:
```diff
@@ -0,0 +1,94 @@
+       Psycopg: http://initd.org/psycopg/
+    """
+
+    if tracer is None:
```
is this check valuable? I don't think we do tracer none checks in other places, and it's a required argument in the call.

toumorokoshi commented on ext/opentelemetry-ext-postgresql/src/opentelemetry/ext/postgresql/__init__.py:
```diff
@@ -0,0 +1,94 @@
+                wrapt.wrap_function_wrapper(self, func, self.add_span)
+            super(TraceCursor, self).__init__(*args, **kwargs)
+
+        def add_span(
```
should this be a protected method? doesn't seem like it should be callable publicly.

toumorokoshi commented on ext/opentelemetry-ext-postgresql/src/opentelemetry/ext/postgresql/__init__.py:
```diff
@@ -0,0 +1,94 @@
+        """Patch Psycopg connect method.
+        """
+        connection = wrapped(*args, **kwargs)
+        connection.cursor_factory = TraceCursor
```
nitpick: the forward reference here took some time for me to grok, I personally prefer the aesthetic of putting this code near the bottom, or at least the "wrapt.wrap_function_wrapper" call so I can see what the final goal is with declaring these inline functions.

hectorhdzg commented:
@toumorokoshi and @Oberon00  I updated this to consume dbapi ext, added basic integration test, this would be super easy to test with functional tests similar to #340 would be happy to add them using similar approach, let me know if you have more feedback

lzchen commented on ext/opentelemetry-ext-psycopg2/README.rst:
```diff
@@ -0,0 +1,25 @@
+    trace_integration(tracer())
+    cnx = psycopg2.connect(database='Database')
+    cursor = cnx.cursor()
+    cursor.execute("INSERT INTO test (testField) VALUES (123)"
```
Missing a ")"

lzchen commented on ext/opentelemetry-ext-psycopg2/README.rst:
```diff
@@ -0,0 +1,25 @@
+    import psycopg2
+    from opentelemetry.trace import tracer
+    from opentelemetry.trace.ext.postgresql import trace_integration
+    trace_integration(tracer())
```
I believe we changed the way we construct `tracer` with [this PR](https://github.com/open-telemetry/opentelemetry-python/pull/301). Might also want to change the other integration examples.

lzchen commented on ext/opentelemetry-ext-psycopg2/README.rst:
```diff
@@ -0,0 +1,25 @@
+.. code:: python
+    import psycopg2
+    from opentelemetry.trace import tracer
+    from opentelemetry.trace.ext.postgresql import trace_integration
```
`from opentelemetry.ext.psycopg2 import trace_integration`

lzchen commented on ext/opentelemetry-ext-psycopg2/tests/test_psycopg2_integration.py:
```diff
@@ -0,0 +1,30 @@
+        with mock.patch("psycopg2.connect"):
+            trace_integration(tracer)
+            cnx = psycopg2.connect(database="test")
+            self.assertIsNotNone(cnx.cursor_factory)
```
I believe the test for `trace_integration` should include to see if the functions are successfully decorated. 
As well, the query functions themselves and the `Cursor` should also have tests.
See [OpenCensus](https://github.com/census-instrumentation/opencensus-python/blob/master/contrib/opencensus-ext-postgresql/tests/test_postgresql_trace.py#L23)

hectorhdzg replied:
I'm only decorating the connect method and this will add the cursor factory, not sure if wrapt will add some "decorated" property, psycopg2 internally will instantiate PsycopgTraceCursor that just call dbapi traced_execution that is already tested, unit tests are hard in here because if the mixed Python/c execution, I believe functional tests would be way more useful and easier to implement, did't want to add them yet until #340 is complete and people are ok with that kind of approach

lzchen commented:
@hectorhdzg  FYI [#360]

toumorokoshi commented:
@hectorhdzg looks good! Can you resolve conflicts (looks like there was some styling stuff on the readmes that conflict with master) and then I can merge?
